sycl/plugins/unified_runtime/ur/adapters/hip/program.cpp

```diff
@@ -26,7 +106,122 @@ ur_result_t ur_program_handle_t_::setBinary(const char *Source, size_t Length) {
+  }
+  // RAII for auto clean-up
+  COMgrDataTCleanUp DataCleanup{ComgrData};
+
+  if (UR_CHECK_ERROR(
+          amd_comgr_set_data(ComgrData, BinarySizeInBytes, Binary))) {
+    std::strcpy(ErrorLog, "Failed to create comgr data set");
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+
+  if (UR_CHECK_ERROR(amd_comgr_set_data_name(ComgrData, "jit_obj.o"))) {
+    std::strcpy(ErrorLog, "Failed to create comgr data set name");
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+
+  if (UR_CHECK_ERROR(amd_comgr_data_set_add(RelocatableData, ComgrData))) {
+    std::strcpy(ErrorLog, "Failed to add data to data set");
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+
+  amd_comgr_action_info_t Action;
+
+  if (UR_CHECK_ERROR(amd_comgr_create_action_info(&Action))) {
+    std::strcpy(ErrorLog, "Failed to create comgr action info");
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+  COMgrActionInfoCleanUp ActionCleanUp{Action};
+
+  std::string ISA = "amdgcn-amd-amdhsa--";
+  hipDeviceProp_t Props;
+  detail::ur::assertion(hipGetDeviceProperties(
+                            &Props, Context->getDevice()->get()) == hipSuccess);
+  ISA += Props.gcnArchName;
+  if (UR_CHECK_ERROR(amd_comgr_action_info_set_isa_name(Action, ISA.data()))) {
+    std::strcpy(ErrorLog, "Failed to set isa ");
+    std::strcat(ErrorLog, ISA.c_str());
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+
+  if (UR_CHECK_ERROR(amd_comgr_action_info_set_logging(Action, true))) {
+    std::strcpy(ErrorLog, "Failed to set logging");
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+
+  amd_comgr_data_set_t Output;
+  if (UR_CHECK_ERROR(amd_comgr_create_data_set(&Output))) {
+    std::strcpy(ErrorLog, "Failed to create dataset for output");
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+  COMgrDataSetTCleanUp OutputDataCleanup{Output};
+
+  if (UR_CHECK_ERROR(
+          amd_comgr_do_action(AMD_COMGR_ACTION_LINK_RELOCATABLE_TO_EXECUTABLE,
+                              Action, RelocatableData, Output))) {
+    getCoMgrBuildLog(Output, ErrorLog, MAX_LOG_SIZE);
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+
+  amd_comgr_data_t binaryData;
+
+  if (UR_CHECK_ERROR(amd_comgr_action_data_get_data(
+          Output, AMD_COMGR_DATA_KIND_EXECUTABLE, 0, &binaryData))) {
+    std::strcpy(ErrorLog, "Failed to get link application");
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+  {
+    COMgrDataTCleanUp binaryDataCleanUp{binaryData};
+
+    size_t binarySize = 0;
+    if (UR_CHECK_ERROR(amd_comgr_get_data(binaryData, &binarySize, NULL))) {
+      std::strcpy(ErrorLog, "Failed to get the executable size");
+      return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+    }
+
+    ExecutableCache.resize(binarySize);
+
+    if (UR_CHECK_ERROR(amd_comgr_get_data(binaryData, &binarySize,
+                                          ExecutableCache.data()))) {
+      std::strcpy(ErrorLog, "Failed to get the executable");
```

Comment:
I did not notice ErrorLog before, it seems like this is the first practical usage of it in any backend.
Is the sycl runtime dealing with "ErrorLog" so that this error message is passed to the user?
Is there a reason to use "ErrorLog" instead of https://github.com/intel/llvm/pull/10953/files#diff-46ec41cb01032cbe27a2f2807b0e9af9621207803463841a1d6d4e43e004fa96R333

They seem to be duplicate error passing mechanisms.

Reply:
> Is there a reason to use "ErrorLog" instead of setErrorMessage?

No, I just reused the buffer for the build log as used for the main path.

```cpp
  // Pass a buffer for error message
  Options[2] = hipJitOptionErrorLogBuffer;
  OptionVals[2] = (void *)ErrorLog;
```

Reply:
I see, does it definitely end up passing the error messages to the runtime? That is the main thing. I just don't know how it works so it might not.